### ISSUE TITLE
Play ► a very simple content in [Audio] mode will crash Shotcut

### ISSUE THREAD

**Issue by chenx77** [mltframework/shotcut#1172]
### A. How to reproduce

Create a simple project, e.g., to have a simple text on a empty canvas as the one in attached
[Study.zip](https://github.com/mltframework/shotcut/files/7428019/Study.zip)

Switch work mode from tab [Editing] to **[Audio]** (yellow oral), then play (**► button** in red circle), as depicted in attached screenshot: ![Screenshot (687)](https://user-images.githubusercontent.com/15191607/139114461-33d0fd03-40ea-4af7-a1dc-f3ccd77fcf33.png)

In most cases the Shotcut crashs at no more than 50 frames (total frame counts 250).

There is no problem to generate a decent mp4 clip, no matter in [Editing] or [Audio] mode. I'm using the default video export: 1920 x 1080, 50 frames/sec, YADIF, Bilinear (good).

### B. Environment

Windows 10 Pro
Processor: Intel(R) Pentium(R) Gold G5500 CPU @ 3.80GHz 3.79 GHz
RAM: 8.00 GB (7.79 GB usable)
System Type: 64-bit operating system, x64-based processor
Shotcut version Win 64, 21.10.17

**Comment by bmatherly** [comment]
I also tried mulitiple times but did not experience a crash.

> Usually I double click the Crash.mlt to open the project.

Is this necessary to cause a crash? Does it crash if you open the file in other ways (like drag to the screen or "file->open")?

**Comment by chenx77** [comment]
> > Usually I double click the Crash.mlt to open the project.
> 
> Is this necessary to cause a crash? Does it crash if you open the file in other ways (like drag to the screen or "file->open")?

It's not necessary. Other ways I open are 1) Lanuch Shoutcut first, then click on Recent Projects, like Crash.mlt; 2) Lanuch Shoutcut, then Open File from Toolbar and load Crash.mlt. No matter in each of 3 ways, [Play ►] will crash 99% times. Here is a screenshot video recorded the process that leads to a crash: https://drive.google.com/file/d/1pstgKE9mtqIKaHE5PP2FaOFIwzXgE633/view?usp=sharing.

If your machine is more powerful than mine, in terms of CPU, RAM etc. (see begining of this thread), or BIO settings, e.g. I have my PC "AESM Service: SGX Enabled", those may explain the difference.

**Comment by ddennedy** [comment]
> And the Qt installed, independently, to my machine may confused the Shotcut?

It will not unless you are trying to launch Shotcut from the tool's project, possibly from our SDK. Or, if any of the Qt SDKs are in any member of the `%PATH%` environment variable. 

Are you able to use Shotcut in general without much crashing? Or, is this example project like the only thing you have ever done with Shotcut?

**Comment by chenx77** [comment]
> 
> 
> > And the Qt installed, independently, to my machine may confused the Shotcut?
> 
> It will not unless you are trying to launch Shotcut from the tool's project, possibly from our SDK. Or, if any of the Qt SDKs are in any member of the `%PATH%` environment variable.

No, when I use Shotcut, it's just the well-built Shotcut, I'm not touch Qt or its Creator at all. I noticed that Shotcut installs Qt5Core.dll into its own file structure (i.e., C:\Program Files\Shotcut\Qt5Core.dll), while Qt/Creator have their own a dozen Qt5Core.dll under Qt path, as a screenshot presented bellow, they shouldn't be mixed up, right?
 
![Screenshot (691)](https://user-images.githubusercontent.com/15191607/139558363-7f5e5a3d-2008-4218-bd2f-2ccae2784c10.png)

> Are you able to use Shotcut in general without much crashing? Or, is this example project like the only thing you have ever done with Shotcut?

Yes, I'm able (using version Win 64, 21.10.17) to generate several decent videos with length of several minutes, without single crash. But when I try to explore more functionalities with layout [Audio], any of those successful projects crash immediately by Play. And that is the reason why I created an as small as possoble project Crash.mlt, to help you to identify the problem.

**Comment by ddennedy** [comment]
Please download the Shotcut SDK for Windows from this page: https://shotcut.org/notes/windowsdev/

It contains a debug build of Shotcut with an integrated crash reporter for use with the MinGW runtime that we use (not MSVC). It is not an online reporter. Rather, when there is a crash, it writes the text file `shotcut.RPT` to the Shotcut program folder. Then, you can upload that here (might need to rename it `.txt`) or copy and paste its contents. You do not need to follow all of the directions on that page to make a custom build. You should be able to simply extract it and run `shotcut.exe` to run the debug build.

**Comment by chenx77** [comment]
> Please download the Shotcut SDK for Windows from this page: https://shotcut.org/notes/windowsdev/
> 
> It contains a debug build of Shotcut with an integrated crash reporter for use with the MinGW runtime that we use (not MSVC).

Do you mean I can run the build "New Version 21.10.31: Markers, Posted on 31 Oct 2021" at right of the page? Or should I build it by Qt Creator/MinGW with the Project?

**Comment by chenx77** [comment]
> 
> 
> > Please download the Shotcut SDK for Windows from this page: https://shotcut.org/notes/windowsdev/
> > It contains a debug build of Shotcut with an integrated crash reporter for use with the MinGW runtime that we use (not MSVC).
> 
> Do you mean I can run the build "New Version 21.10.31: Markers, Posted on 31 Oct 2021" at right of the page? Or should I build it by Qt Creator/MinGW with the Project?

OK, now I see the build inside the Zip package. Sorry to raise question too quick.

**Comment by chenx77** [comment]
> Please download the Shotcut SDK for Windows from this page: https://shotcut.org/notes/windowsdev/
> 
> It contains a debug build of Shotcut with an integrated crash reporter for use with the MinGW runtime that we use (not MSVC). It is not an online reporter. Rather, when there is a crash, it writes the text file `shotcut.RPT` to the Shotcut program folder. Then, you can upload that here (might need to rename it `.txt`) or copy and paste its contents. You do not need to follow all of the directions on that page to make a custom build. You should be able to simply extract it and run `shotcut.exe` to run the debug build.

Here you are, the Report file [shotcut.txt](https://github.com/mltframework/shotcut/files/7452399/shotcut.txt). The build is a bit harder to crash, but I still manage to crash it by shifting between [Audio] & [Editing], of course by [►].

**Comment by ddennedy** [comment]
Your log shows a crash inside Qt when Shotcut is updating a text label in its Audio Loudness scope. However, what is more interesting is that the function call backtrace between these two points shows a number of function calls within a `UIAutomationCore.dll`. I have never seen that before, and it is strange. This is some DLL that is specific to your computer environment that has hooked itself into the process. These kinds of system hooks are not supported as they can interfere with normal operations and introduce unknown variables. So, I am closing this as will not fix. Instead, you can try to fix your operating system environment or reinstall it. You can learn about this DLL here:
https://en.wikipedia.org/wiki/Microsoft_UI_Automation

**Comment by bmatherly** [comment]
From this trace, I notice that QLable::setText() is called outside of the GUI thread.

```
> Qt5Widgets.dll!QLabel::setText
> shotcut.exe!AudioLoudnessScopeWidget::refreshScope
> shotcut.exe!ScopeWidget::refreshInThread
```

I do not know if that is causing this specific problem, but i changed it here to use the GUI thread:
https://github.com/mltframework/shotcut/commit/6c8264b7ec007c325e642b44d07aa0e16a0019d0

**Comment by chenx77** [comment]
> 
> 
> From this trace, I notice that QLable::setText() is called outside of the GUI thread.
> 
> ```
> > Qt5Widgets.dll!QLabel::setText
> > shotcut.exe!**AudioLoudness**ScopeWidget::refreshScope
> > shotcut.exe!ScopeWidget::refreshInThread

In deed! This can be proved by a test: whenever the [Audio Loudness] widget is switch on (yellow rectangle), the crash is inevitable, no matter in [Audio] or [Editing] (red oval) layout.
![Screenshot (698)](https://user-images.githubusercontent.com/15191607/139846958-ed85a0d7-09fa-4aed-b097-61e3ab9b2824.png)

**Comment by ddennedy** [comment]
You can test this change with the nightly build here: https://github.com/mltframework/shotcut/actions/runs/1411352735

**Comment by chenx77** [comment]
> You can test this change with the nightly build here: https://github.com/mltframework/shotcut/actions/runs/1411352735

Thank you! The reported crash no longer exist, the issue has been fixed by **build 21.11.02**. The original report didn't hit the major point, the problem is NOT at mode [Audio] or [Editing], but **relys on if widget [Audio Loudness] is presented or not**.

Thank you again!

**Comment by ddennedy** [comment]
Thank you for testing! This will be in the next release due by end of year (there is a possibility that 21.11 will be converted to 21.12 to be released in December instead of November).
